lib/Value/Matrix.pm

```diff
@@ -416,58 +451,105 @@ sub isZero {
 	return 1;
 }
 
-#
-#  See if the matrix is triangular, diagonal, symmetric, orthogonal
-#
+=head3 C<isUpperTriangular>
+
+Check if a Matrix is upper triangular (for degree > 2, applies to frontal slice matrices)
+
+=cut
 
 sub isUpperTriangular {
 	my $self = shift;
 	my @d    = $self->dimensions;
 	return 1 if scalar(@d) == 1;
-	return 0 if scalar(@d) > 2;
-	for my $i (2 .. $d[0]) {
-		for my $j (1 .. ($i - 1 < $d[1] ? $i - 1 : $d[1])) {
-			return 0 unless $self->element($i, $j) == 0;
+	if (scalar(@d) == 2) {
+		for my $i (2 .. $d[0]) {
+			for my $j (1 .. ($i - 1 < $d[1] ? $i - 1 : $d[1])) {
+				return 0 unless $self->element($i, $j) == 0;
+			}
+		}
+	} else {
+		for my $row (@{ $self->{data} }) {
+			if (!$row->isUpperTriangular) {
+				return 0;
+			}
 		}
 	}
 	return 1;
 }
 
+=head3 C<isLowerTriangular>
+
+Check if a Matrix is lower triangular (for degree > 2, applies to frontal slice matrices)
+
+=cut
+
 sub isLowerTriangular {
 	my $self = shift;
 	my @d    = $self->dimensions;
 	if (scalar(@d) == 1) {
 		for ((@{ $self->{data} })[ 1 .. $#{ $self->{data} } ]) {
 			return 0 unless $_ == 0;
 		}
-	}
-	return 0 if scalar(@d) > 2;
-	for my $i (1 .. $d[0] - 1) {
-		for my $j ($i + 1 .. $d[1]) {
-			return 0 unless $self->element($i, $j) == 0;
+	} elsif (scalar(@d) == 2) {
+		for my $i (1 .. $d[0]) {
+			for my $j ($i + 1 .. $d[1]) {
+				return 0 unless $self->element($i, $j) == 0;
+			}
+		}
+	} else {
+		for my $row (@{ $self->{data} }) {
+			if (!$row->isLowerTriangular) {
+				return 0;
+			}
```

Comment:
Same comment here as above.  I see this also applies to several places below as well.